Add support for extension tiers

jcfr opened the pull request:
Update Slicer to extension tiers described in https://discourse.slicer.org/t/introduction-of-tiers-for-slicer-extensions/34870

Summary:
* Add `v1.0.1` of extensions catalog entry schema (add `tier` optional metadata)
* Misc test fixes and stylistic changes
* Add extension tier support _work-in-progress_ :hourglass_flowing_sand: 

Next:
* [x] Update https://github.com/girder/slicer_package_manager (server & client) to support `tier`
  * https://github.com/girder/slicer_package_manager/pull/124
* [ ] Update frontend https://github.com/KitwareMedical/slicer-extensions-webapp to report the extension tier and allow filtering by tier
* [ ] Update Slicer extension manager to de-emphasize extensions below tier 3. Either hide them by default or show in a separate section below higher-tier extensions.
* [ ] Update https://github.com/Slicer/ExtensionsIndex/blob/main/.github/PULL_REQUEST_TEMPLATE.md based on tier
* [ ] Update Slicer extension documentation 
* [x] Update `.json` files in extension index based on `v1.0.1` schema

jcfr committed:
ENH: Add extension tier support

Introduce `TIER` as optional metadata with default "1" and propagate
through spec, generation, extraction, packaging, and upload. Update
templates and tests accordingly.

jcfr commented:
@jamesobutler @sjh26 @lassoan @pieper  Do not hesitate to provide feedback regarding the _Next_ steps documented above :pray:

lassoan commented on Extensions/CMake/Testing/extension_description_with_depends.s4ext:
```diff
@@ -34,6 +34,10 @@ iconurl https://www.slicer.org/w/images/6/64/SlicerToKiwiExporterLogo.png
 status Development
 
+# Extension tier: 1, 3, or 5
+# See https://discourse.slicer.org/t/introduction-of-tiers-for-slicer-extensions/34870
```
Refer to readthedocs documentation page instead.

jcfr committed:
ENH: Ensure extension archive include the expected metadata

Rewrite the ".s4ext" with metadata extracted from ".json" and
ensure the expected metadata are associated with the archive.

lassoan commented on Extensions/CMake/SlicerExtensionDescriptionSpec.cmake:
```diff
@@ -124,6 +124,11 @@ _define_metadata("STATUS"
+_define_metadata("TIER"
+  "Extension tier"
+  0
+  "5")
```
```suggestion
  "1")
```

lassoan commented on Extensions/CMake/SlicerFunctionExtractExtensionDescription.cmake:
```diff
@@ -113,6 +113,7 @@ endfunction()
 #  <var-prefix>_EXT_BUILD_DEPENDENCIES - list of Slicer extensions required at build-time
 #  <var-prefix>_EXT_CATEGORY - category
 #  <var-prefix>_EXT_ENABLED - indicate if the extension should be enabled after its installation (default is 1)
+#  <var-prefix>_EXT_TIER - extension tier (default is 5)
```
```suggestion
#  <var-prefix>_EXT_TIER - extension tier (default is 1)
```

lassoan commented on Extensions/CMake/SlicerFunctionExtractExtensionDescription.cmake:
```diff
@@ -149,11 +150,13 @@ function(slicerFunctionExtractExtensionDescriptionFromJson)
   set(BUILD_SUBDIRECTORY_DEFAULT ".")
   set(BUILD_DEPENDENCIES_DEFAULT "")
   set(ENABLED_DEFAULT "1")
+  set(TIER_DEFAULT "5")
```
```suggestion
  set(TIER_DEFAULT "1")
```

lassoan commented on Extensions/CMake/SlicerFunctionExtractExtensionDescription.cmake:
```diff
@@ -356,6 +360,7 @@ function(slicer_extract_extension_description_from_json_test)
   set(expected_BUILD_SUBDIRECTORY ".")
   set(expected_BUILD_DEPENDENCIES "")
   set(expected_ENABLED "1")
+  set(expected_TIER "5")
```
```suggestion
  set(expected_TIER "1")
```

lassoan commented on Extensions/CMake/SlicerFunctionExtractExtensionDescription.cmake:
```diff
@@ -380,7 +385,8 @@ function(slicer_extract_extension_description_from_json_test)
     \"scm_url\": \"${expected_SCM_URL}\",
-    \"enabled\": false
+    \"enabled\": false,
+    \"tier\": 1,
```
```suggestion
    \"tier\": 5,
```

lassoan commented on Extensions/CMake/SlicerFunctionExtractExtensionDescription.cmake:
```diff
@@ -392,6 +398,7 @@ function(slicer_extract_extension_description_from_json_test)
   set(expected_BUILD_SUBDIRECTORY "inner-build")
   set(expected_BUILD_DEPENDENCIES Foo Bar)
   set(expected_ENABLED "0")
+  set(expected_TIER "1")
```
```suggestion
  set(expected_TIER "5")
```

lassoan commented on Extensions/CMake/SlicerFunctionGenerateExtensionDescription.cmake:
```diff
@@ -168,6 +169,7 @@ This is a line of text.<br>And another one."
   # Generate description file of an extension *without* dependencies
   slicerFunctionGenerateExtensionDescription(
     ${common_args}
+    EXTENSION_TIER "1"
```
```suggestion
    EXTENSION_TIER "5"
```

lassoan commented on Extensions/CMake/SlicerFunctionGenerateExtensionDescription.cmake:
```diff
@@ -188,6 +190,7 @@ This is a line of text.<br>And another one."
   # where EXTENSION_DEPENDS is a space separated string
   slicerFunctionGenerateExtensionDescription(
     ${common_args}
+    EXTENSION_TIER "5"
```
```suggestion
    EXTENSION_TIER "1"
```

lassoan commented on Extensions/CMake/Testing/extension_description_with_depends.s4ext:
```diff
@@ -34,6 +34,10 @@ iconurl https://www.slicer.org/w/images/6/64/SlicerToKiwiExporterLogo.png
 
+# Extension tier: 1, 3, or 5
+# See https://slicer.readthedocs.io/en/latest/developer_guide/extensions.html
+tier 5
```
```suggestion
tier 1
```

lassoan commented on Extensions/CMake/Testing/extension_description_without_depends.s4ext:
```diff
@@ -34,6 +34,10 @@ iconurl https://www.slicer.org/w/images/6/64/SlicerToKiwiExporterLogo.png
 
+# Extension tier: 1, 3, or 5
+# See https://slicer.readthedocs.io/en/latest/developer_guide/extensions.html
+tier 1
```
```suggestion
tier 5
```

lassoan commented on Schemas/slicer-extension-catalog-entry-schema-v1.0.1.json:
```diff
@@ -0,0 +1,65 @@
+      "type": "integer",
+      "title": "Tier of the extension.",
+      "enum": [1, 3, 5],
+      "default": 5
```
```suggestion
      "default": 1
```

lassoan commented on Schemas/slicer-extension-catalog-entry-schema-v1.0.1.json:
```diff
@@ -0,0 +1,65 @@
+    "tier": {
+      "$id": "#tier",
+      "type": "integer",
+      "title": "Tier of the extension.",
```
```suggestion
      "title": "Tier of the extension. Higher value means higher level of support.",
```

lassoan reviewed (commented):
Set default level to 1, according to recent discussions to change meaning of tier to "higher value means higher level of support"
